features/268-next-cf/api.md

```diff
@@ -0,0 +1,625 @@
+- Path of `http-handler` and http `function` are exact match
+- Path of `http-service` are matched with prefix
+- Longer path will be matched before the shorter ones
+- If none of the registered path is matched, return function not found error
+
+Static assets share another domain, so routing is done separately, the rules are similar:
+- Path are matched with prefix
+- Longer path will be matched before shorter ones
+- If none of the registered path is matched, return asset not found error
+
+For example, with the following skycli configuration:
+
+```yaml
+cf:
+  functionABC:
+    type: http-handler
+    path: /functionABC
+    # other config
+  function-server:
+    type: http-service
+    path: /function
+    # other config
+  api-server:
+    type: http-service
+    path: /api
+    # other config
+
+static:
+  - src: index.html
+    path: /
+  - src: asset
+    path: /static
+```
+
+- `https://myapp.api.skygear.io/function/ABC` -> `function-server`, forwarded path `/ABC`
+- `https://myapp.api.skygear.io/functionABC` -> `functionABC`
+- `https://myapp.api.skygear.io/functionABCD` -> `function-server`, forwarded path `ABCD`
+- `https://myapp.api.skygear.io/function/` -> `function-server`, forwarded path `/`
+- `https://myapp.api.skygear.io/function` -> `function-server`, forwarded path ``
+- `https://myapp.api.skygear.io/api/function` -> `api-server`, forwarded path `/function`
+- `https://myapp.api.skygear.io/api/` -> `api-server`, forwarded path `/`
+- `https://myapp.api.skygear.io/api` -> `api-server`, forwarded path ``
+- `https://myapp.api.skygear.io/static` -> function not found
+- `https://myapp.api.skygear.io/` -> function not found
+
+- `https://myapp.skygear.io/static/abc/cde.jpg` -> static assets with path `/abc/cde.jpg` in `asset`
+- `https://myapp.skygear.io/static/abc` -> static assets with path `/abc` in `asset`
+- `https://myapp.skygear.io/` -> static assets `index.html`
+- `https://myapp.skygear.io/any/other/path` -> static assets `index.html`
+
+# Example Configuration
+
+## Mixing functions with micro-service
+
+```yaml
+cf:
+  function1:
+    type: http-handler
+    path: /function1
+    env: node
+    src: /functions/1
+    permission:
+      - name: key_required
+  function2:
+    type: http-handler
+    path: /function2
+    env: node
+    src: /functions/2
+    permission:
+      - name: key_required
+      - name: user_required
+  api-server:
+    type: http-service
+    path: /api
+    env: node
+    src: js
+    secrets:
+      - DATABASE_URL
+    permission:
+      - name: key_required
+      - name: user_required
+```
+
+There are three separate function configurations here, the `api-server` is a `http-service` so will match all `/api/*` request.
+
+## Typical SPA with api server
+
+```yaml
+cf:
+  api-server:
+    type: http-service
+    path: /api
+    env: golang
+    src: api
+    secrets:
+      - DATABASE_URL
+
+static:
+  - src: build/index.html
+    path: /
+  - src: build/asset
+    path: /static
+```
```

Comment:
Because of some files (like favicon.ico, manifest.json) are stored in root folder by default (for example I use `react-create-app` and use `npm run build` to build production app), may I know does it allow to have duplicate `path`? like

```
static:
  - src: build/index.html
    path: /
  - src: build
    path: /
  - src: build/static
    path: /static
``` 

or (`static` folder is upload automatically)

```
  - src: build/index.html
    path: /
  - src: build
    path: /
```

Reply:
```
static:
  - src: build/index.html
    path: /
  - src: build
    path: /
  - src: build/static
    path: /static
```

The above config is not a valid one.

And because my proposal does not follow a directory structure of common build tools, I expect the developer needs to update the build tools config, or rearrange the file location after being built.

Reply:
I have to be honest I didn't look/think clearly into it, I was just triggered by the statement "because my proposal does not follow a directory structure of common build tools"

It sounds like something might seriously affect the user's experience?

Reply:
I suddenly realise we can add a little config so that it can fit common build tools and fulfil my original proposal.

```
static:
  - src: build/index.html
    path: /
  - src: build
    exclude:
       - index.html <- this will exclude the file build/index.html
    path: /static
```

With the above configuration, everything can be put in the `build` directory, while the server won't serve the path `https://myapp.skygeario.com/static/index.html`.

(Though, actually, it is not a really big deal to also serve the path `https://myapp.skygeario.com/static/index.html`)

Reply:
@Steven-Chan Great. Another idea is what if we have longest prefix match on src too? In that case we don't need users to specify exclude?

(Or yes, maybe it is not a big deal to also serve the path of `/static/index.html`)